// @bovine/mucked-up-components/src/ActorHeader/index.tsx
import type { ActorHeaderInfo } from "@bovine/muck-out-types";

import {
  ActorHeaderIdentifierDiv,
  ActorHeaderLink,
  ActorHeaderNameDiv,
} from "./styles";
import { ActorAvatar } from "../ActorAvatar";

type ActorHeaderProps = {
  actorInfo: ActorHeaderInfo;
  style?: object;
};

function transformIdentifier(identifier: string): string {
  return identifier.replace("acct:", "@");
}

function ActorHeaderName({ actorInfo }: ActorHeaderProps) {
  const name = actorInfo.name || "- unknown -";

  if (actorInfo.htmlUrl) {
    return (
      <ActorHeaderNameDiv>
        <ActorHeaderLink href={actorInfo.htmlUrl}>{name}</ActorHeaderLink>
      </ActorHeaderNameDiv>
    );
  }

  return <ActorHeaderNameDiv>{name}</ActorHeaderNameDiv>;
}

function ActorHeaderIdentifier({ actorInfo }: ActorHeaderProps) {
  return (
    <ActorHeaderIdentifierDiv>
      <ActorHeaderLink href={actorInfo.id}>
        {transformIdentifier(actorInfo.identifier)}
      </ActorHeaderLink>
    </ActorHeaderIdentifierDiv>
  );
}

function ActorHeader({ actorInfo, style }: ActorHeaderProps) {
  return (
    <header style={{ ...style, display: "flex" }}>
      <ActorAvatar actorInfo={actorInfo} />
      <span>
        <ActorHeaderName actorInfo={actorInfo} />
        <ActorHeaderIdentifier actorInfo={actorInfo} />
      </span>
    </header>
  );
}

export { ActorHeader, type ActorHeaderProps };
