// @bovine/mucked-up-components/src/ActorAvatarList/ActorAvatarList.stories.tsx
import type { Meta, StoryObj } from "@storybook/react-vite";

import { ActorAvatarList } from ".";

const actor = {
  id: "http://abel/actor/AFKb0cQunSBv1fC7sWbQYg",
  avatarUrl: "https://dev.bovine.social/assets/bull-horns.png",
  name: "The kitty",
  identifier: "acct:kitty@abel",
  htmlUrl: "http://abel/@kitty",
};

const actor2 = {
  id: "https://mymath.rocks/endpoints/SYn3cl_N4HAPfPHgo2x37XunLEmhV9LnxCggcYwyec0",
  avatarUrl: "https://mymath.rocks/static/profile.png",
  name: "Helge",
  identifier: "acct:helge@mymath.rocks",
  htmlUrl: "https://mymath.rocks/timeline/",
};

const meta = {
  component: ActorAvatarList,
} satisfies Meta<typeof ActorAvatarList>;

export default meta;
type Story = StoryObj<typeof meta>;

export const FullInformation: Story = {
  args: {
    actors: [actor, actor2],
  },
};

export const DisplaySmall: Story = {
  args: {
    actors: [actor, actor2],
    avatarStyle: { width: "15px", height: "15px", marginRight: "2px" },
  },
};

export const DisplayOverlap: Story = {
  args: {
    actors: [actor, actor2, { ...actor, id: actor.id + "x" }],
    avatarStyle: { marginRight: "-20px" },
  },
};
