// @bovine/mucked-up-components/src/ObjectMeta/index.tsx
import type { ObjectMetaInfo } from "@bovine/muck-out-types";

import dayjs from "dayjs";
import relativeTime from "dayjs/plugin/relativeTime";
import { LinkNoUnderline } from "./styles";

dayjs.extend(relativeTime);

type ObjectMetaProps = {
  info: ObjectMetaInfo;
};

function formatDate(date: string) {
  const day = dayjs(date);
  return day.fromNow();
}

function DisplayDate({ info }: ObjectMetaProps) {
  if (info.updated) {
    return <>edited, {formatDate(info.updated)}</>;
  }
  if (info.published) {
    return <>{formatDate(info.published)}</>;
  }
  return <>post</>;
}

function HtmlUrl({ info }: ObjectMetaProps) {
  if (!info.htmlUrl) {
    return <></>;
  }
  return (
    <>
      , <LinkNoUnderline href={info.htmlUrl}>www</LinkNoUnderline>
    </>
  );
}

function ObjectMeta({ info }: ObjectMetaProps) {
  return (
    <div>
      <LinkNoUnderline href={info.id}>
        <DisplayDate info={info} />
      </LinkNoUnderline>
      <HtmlUrl info={info} />
    </div>
  );
}

export { ObjectMeta, type ObjectMetaProps };
