<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" /><meta name="viewport" content="width=device-width, initial-scale=1" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>Installation &mdash; genal v0.8 documentation</title>
      <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
      <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />
  <!--[if lt IE 9]>
    <script src="_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
        <script src="_static/documentation_options.js?v=b326c068"></script>
        <script src="_static/doctools.js?v=9a2dae69"></script>
        <script src="_static/sphinx_highlight.js?v=dc90522c"></script>
    <script src="_static/js/theme.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="The Geno class" href="modules.html" />
    <link rel="prev" title="genal: A Python Toolkit for Genetic Risk Scoring and Mendelian Randomization" href="index.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >

          
          
          <a href="index.html" class="icon icon-home">
            genal
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <ul class="current">
<li class="toctree-l1"><a class="reference internal" href="index.html">Home</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="#tutorial">Tutorial</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#table-of-contents">Table of contents</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#data-loading">Data loading</a></li>
<li class="toctree-l3"><a class="reference internal" href="#data-preprocessing">Data preprocessing</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#clumping">Clumping</a></li>
<li class="toctree-l2"><a class="reference internal" href="#polygenic-risk-scoring">Polygenic Risk Scoring</a></li>
<li class="toctree-l2"><a class="reference internal" href="#mendelian-randomization">Mendelian Randomization</a></li>
<li class="toctree-l2"><a class="reference internal" href="#snp-association-testing">SNP-association testing</a></li>
<li class="toctree-l2"><a class="reference internal" href="#lifting">Lifting</a></li>
<li class="toctree-l2"><a class="reference internal" href="#gwas-catalog">GWAS Catalog</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="modules.html">The Geno class</a></li>
<li class="toctree-l1"><a class="reference internal" href="modules.html#main-functions">Main functions</a></li>
<li class="toctree-l1"><a class="reference internal" href="api.html">API</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">genal</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="index.html" class="icon icon-home" aria-label="Home"></a></li>
      <li class="breadcrumb-item active">Installation</li>
      <li class="wy-breadcrumbs-aside">
            <a href="_sources/introduction.rst.txt" rel="nofollow"> View page source</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section id="installation">
<h1>Installation<a class="headerlink" href="#installation" title="Link to this heading">¶</a></h1>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p><strong>Optional</strong>: It is recommended to create a new environment to avoid dependencies conflicts. Here, we create a new conda environment called ‘genal’.</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>conda<span class="w"> </span>create<span class="w"> </span>--name<span class="w"> </span>genal<span class="w"> </span><span class="nv">python</span><span class="o">=</span><span class="m">3</span>.11
conda<span class="w"> </span>activate<span class="w"> </span>genal
</pre></div>
</div>
</div>
<p>The genal package requires Python 3.11. Download and install it with pip:</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>pip<span class="w"> </span>install<span class="w"> </span>genal-python
</pre></div>
</div>
<p>And import it in a python environment with:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">genal</span>
</pre></div>
</div>
<p>The main genal functionalities require a working installation of PLINK v1.9 (and not 2.0 as certain functionalities have not been updated yet) that can be downloaded here: <a class="reference external" href="https://www.cog-genomics.org/plink/">https://www.cog-genomics.org/plink/</a>
Once downloaded, the path to the plink 1.9 executable should be set with:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">genal</span><span class="o">.</span><span class="n">set_plink</span><span class="p">(</span><span class="n">path</span><span class="o">=</span><span class="s2">&quot;/path/to/plink/executable/file&quot;</span><span class="p">)</span>
</pre></div>
</div>
</section>
<section id="tutorial">
<h1>Tutorial<a class="headerlink" href="#tutorial" title="Link to this heading">¶</a></h1>
<p>For the purpose of this tutorial, we are going to build a PRS of systolic blood pressure (SBP) and investigate the genetically-determined effect of SBP on the risk of stroke. We will use both summary statistics from Genome-Wide Association Studies (GWAS) and individual-level data from the UK Biobank as our test population. We are going to go through the following steps:</p>
<section id="table-of-contents">
<h2>Table of contents<a class="headerlink" href="#table-of-contents" title="Link to this heading">¶</a></h2>
<ol class="loweralpha simple">
<li><p><a class="reference internal" href="#data-loading">Data loading</a></p></li>
<li><p><a class="reference internal" href="#data-preprocessing">Data preprocessing</a></p></li>
<li><p><a class="reference internal" href="#clumping">Clumping</a></p></li>
<li><p><a class="reference internal" href="#polygenic-risk-scoring">Polygenic Risk Scoring</a></p></li>
<li><p><a class="reference internal" href="#mendelian-randomization">Mendelian Randomization</a></p></li>
<li><p><a class="reference internal" href="#snp-association-testing">SNP-association testing</a></p></li>
<li><p><a class="reference internal" href="#lifting">Lifting</a></p></li>
<li><p><a class="reference internal" href="#gwas-catalog">GWAS Catalog</a></p></li>
</ol>
<section id="data-loading">
<h3>Data loading<a class="headerlink" href="#data-loading" title="Link to this heading">¶</a></h3>
<p>We start this tutorial with publicly available summary statistics data from a large GWAS of systolic blood pressure (<a class="reference external" href="https://www.nature.com/articles/s41588-018-0205-x">https://www.nature.com/articles/s41588-018-0205-x</a>). After downloading and unzipping the summary statistics, we load the data into a pandas dataframe:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>
<span class="n">sbp_gwas</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="s2">&quot;Evangelou_30224653_SBP.txt&quot;</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="s2">&quot; &quot;</span><span class="p">)</span>
</pre></div>
</div>
<p>The loaded dataframe:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">sbp_gwas</span><span class="o">.</span><span class="n">head</span><span class="p">(</span><span class="mi">5</span><span class="p">)</span>
<span class="go">         MarkerName Allele1 Allele2   Freq1  Effect  StdErr         P  TotalSampleSize  N_effective</span>
<span class="go">0  10:100000625:SNP       a       g  0.5660  0.0523  0.0303  0.083940           738170       736847</span>
<span class="go">1  10:100000645:SNP       a       c  0.7936  0.0200  0.0372  0.591100           738168       735018</span>
<span class="go">2  10:100003242:SNP       t       g  0.8831  0.1417  0.0469  0.002526           738168       733070</span>
<span class="go">3  10:100003304:SNP       a       g  0.9609  0.0245  0.0838  0.769800           737054       663809</span>
<span class="go">4  10:100003785:SNP       t       c  0.6406 -0.0680  0.0313  0.029870           738169       735681</span>
</pre></div>
</div>
<p>We can now load this data into a <a class="reference internal" href="modules.html#genal.Geno" title="genal.Geno"><code class="xref py py-class docutils literal notranslate"><span class="pre">Geno</span></code></a> object. The <a class="reference internal" href="modules.html#genal.Geno" title="genal.Geno"><code class="xref py py-class docutils literal notranslate"><span class="pre">Geno</span></code></a> class is the central piece of the package. It is designed to store Single Nucleotide Polymorphisms (SNP) data and make it easy to preprocess and clean.</p>
<p>The <a class="reference internal" href="modules.html#genal.Geno" title="genal.Geno"><code class="xref py py-class docutils literal notranslate"><span class="pre">Geno</span></code></a> takes as input a pandas dataframe where each row corresponds to a SNP, with columns describing the position and possibly the effect of the SNP for the given trait (SBP in our case). The following arguments can be passed to specify the column names:</p>
<ul class="simple">
<li><p><strong>CHR</strong>: Column name for chromosome. Defaults to “CHR”.</p></li>
<li><p><strong>POS</strong>: Column name for genomic position. Defaults to “POS”.</p></li>
<li><p><strong>SNP</strong>: Column name for SNP identifier (rsid). Defaults to “SNP”.</p></li>
<li><p><strong>EA</strong>: Column name for effect allele. Defaults to “EA”.</p></li>
<li><p><strong>NEA</strong>: Column name for non-effect allele. Defaults to “NEA”.</p></li>
<li><p><strong>BETA</strong>: Column name for effect estimate. Defaults to “BETA”.</p></li>
<li><p><strong>SE</strong>: Column name for effect standard error. Defaults to “SE”.</p></li>
<li><p><strong>P</strong>: Column name for effect p-value. Defaults to “P”.</p></li>
<li><p><strong>EAF</strong>: Column name for effect allele frequency. Defaults to “EAF”.</p></li>
</ul>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>You do not need all columns to move forward, as not all columns are required by every function. Additionally, some columns can be imputed as we will see in the next paragraph.</p>
</div>
<p>In our case, and after inspecting the dataframe, we must first extract the chromosome and position information from the “MarkerName” column into two new columns “CHR” and “POS”:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">sbp_gwas</span><span class="p">[[</span><span class="s2">&quot;CHR&quot;</span><span class="p">,</span> <span class="s2">&quot;POS&quot;</span><span class="p">,</span> <span class="s2">&quot;Filler&quot;</span><span class="p">]]</span> <span class="o">=</span> <span class="n">sbp_gwas</span><span class="p">[</span><span class="s2">&quot;MarkerName&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">str</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;:&quot;</span><span class="p">,</span> <span class="n">expand</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
</pre></div>
</div>
<p>The resulting dataframe now has separate columns for the chromosome and genomic position:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">sbp_gwas</span><span class="o">.</span><span class="n">head</span><span class="p">(</span><span class="mi">5</span><span class="p">)</span>
<span class="go">         MarkerName Allele1 Allele2   Freq1  Effect  StdErr         P  TotalSampleSize  N_effective CHR        POS Filler</span>
<span class="go">0  10:100000625:SNP       a       g  0.5660  0.0523  0.0303  0.083940           738170       736847  10  100000625    SNP</span>
<span class="go">1  10:100000645:SNP       a       c  0.7936  0.0200  0.0372  0.591100           738168       735018  10  100000645    SNP</span>
<span class="go">2  10:100003242:SNP       t       g  0.8831  0.1417  0.0469  0.002526           738168       733070  10  100003242    SNP</span>
<span class="go">3  10:100003304:SNP       a       g  0.9609  0.0245  0.0838  0.769800           737054       663809  10  100003304    SNP</span>
<span class="go">4  10:100003785:SNP       t       c  0.6406 -0.0680  0.0313  0.029870           738169       735681  10  100003785    SNP</span>
</pre></div>
</div>
<p>and it can now be loaded into a <a class="reference internal" href="modules.html#genal.Geno" title="genal.Geno"><code class="xref py py-class docutils literal notranslate"><span class="pre">Geno</span></code></a> object:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">genal</span>
<span class="n">SBP_Geno</span> <span class="o">=</span> <span class="n">genal</span><span class="o">.</span><span class="n">Geno</span><span class="p">(</span><span class="n">sbp_gwas</span><span class="p">,</span> <span class="n">CHR</span><span class="o">=</span><span class="s2">&quot;CHR&quot;</span><span class="p">,</span> <span class="n">POS</span><span class="o">=</span><span class="s2">&quot;POS&quot;</span><span class="p">,</span> <span class="n">EA</span><span class="o">=</span><span class="s2">&quot;Allele1&quot;</span><span class="p">,</span> <span class="n">NEA</span><span class="o">=</span><span class="s2">&quot;Allele2&quot;</span><span class="p">,</span> <span class="n">BETA</span><span class="o">=</span><span class="s2">&quot;Effect&quot;</span><span class="p">,</span> <span class="n">SE</span><span class="o">=</span><span class="s2">&quot;StdErr&quot;</span><span class="p">,</span> <span class="n">P</span><span class="o">=</span><span class="s2">&quot;P&quot;</span><span class="p">,</span> <span class="n">EAF</span><span class="o">=</span><span class="s2">&quot;Freq1&quot;</span><span class="p">,</span> <span class="n">keep_columns</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
</pre></div>
</div>
<p>The last argument (<code class="docutils literal notranslate"><span class="pre">keep_columns</span> <span class="pre">=</span> <span class="pre">False</span></code>) indicates that we do not wish to keep the other (non-main) columns in the dataframe.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>Make sure to read the readme file usually provided with the summary statistics to identify the correct columns. It is particularly important to correctly identify the allele that represents the effect allele.</p>
</div>
</section>
<section id="data-preprocessing">
<h3>Data preprocessing<a class="headerlink" href="#data-preprocessing" title="Link to this heading">¶</a></h3>
<p>Now that we have loaded the data into a <a class="reference internal" href="modules.html#genal.Geno" title="genal.Geno"><code class="xref py py-class docutils literal notranslate"><span class="pre">Geno</span></code></a> instance, we can begin cleaning and formatting it. Methods such as Polygenic Risk Scoring or Mendelian Randomization require the SNP data to be in a specific format. Additionally, raw summary statistics can sometimes contain missing or invalid values that need to be handled. Some columns may be missing from the data (such as the SNP rsid column or the non-effect allele column), and these columns can be created based on existing ones and a reference panel.</p>
<p>Genal can run all the basic cleaning and preprocessing steps in one command:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">SBP_Geno</span><span class="o">.</span><span class="n">preprocess_data</span><span class="p">(</span><span class="n">preprocessing</span><span class="o">=</span><span class="s1">&#39;Fill_delete&#39;</span><span class="p">)</span>
</pre></div>
</div>
<p>The <code class="docutils literal notranslate"><span class="pre">preprocessing</span></code> argument specifies the global level of preprocessing applied to the data:</p>
<ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">preprocessing</span> <span class="pre">=</span> <span class="pre">'None'</span></code>: The data won’t be modified.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">preprocessing</span> <span class="pre">=</span> <span class="pre">'Fill'</span></code>: Missing columns will be added based on reference data and invalid values set to NaN, but no rows will be deleted.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">preprocessing</span> <span class="pre">=</span> <span class="pre">'Fill_delete'</span></code>: Missing columns will be added, and all rows containing missing, duplicated, or invalid values will be deleted. This option is recommended before running genetic methods.</p></li>
</ul>
<p>Defaults to <code class="docutils literal notranslate"><span class="pre">'Fill'</span></code>.</p>
<p>By default, and depending on the global preprocessing level (<code class="docutils literal notranslate"><span class="pre">'None'</span></code>, <code class="docutils literal notranslate"><span class="pre">'Fill'</span></code>, <code class="docutils literal notranslate"><span class="pre">'Fill_delete'</span></code>) chosen, the <a class="reference internal" href="modules.html#id0" title="genal.Geno.preprocess_data"><code class="xref py py-meth docutils literal notranslate"><span class="pre">preprocess_data()</span></code></a> method of <a class="reference internal" href="modules.html#genal.Geno" title="genal.Geno"><code class="xref py py-class docutils literal notranslate"><span class="pre">Geno</span></code></a> will run the following checks:</p>
<ul class="simple">
<li><p>Ensure the <code class="docutils literal notranslate"><span class="pre">CHR</span></code> (chromosome) and <code class="docutils literal notranslate"><span class="pre">POS</span></code> (genomic position) columns are integers.</p></li>
<li><p>Ensure the <code class="docutils literal notranslate"><span class="pre">EA</span></code> (effect allele) and <code class="docutils literal notranslate"><span class="pre">NEA</span></code> (non-effect allele) columns are uppercase characters containing A, T, C, G letters. Multiallelic values are set to NaN.</p></li>
<li><p>Validate the <code class="docutils literal notranslate"><span class="pre">P</span></code> (p-value) column for proper values.</p></li>
<li><p>Check for no duplicated SNPs based on rsid.</p></li>
<li><p>Determine if the <code class="docutils literal notranslate"><span class="pre">BETA</span></code> (effect) column contains beta estimates or odds ratios, and log-transform odds ratios if necessary.</p></li>
<li><p>Create <code class="docutils literal notranslate"><span class="pre">SNP</span></code> column using a reference panel if <code class="docutils literal notranslate"><span class="pre">CHR</span></code> and <code class="docutils literal notranslate"><span class="pre">POS</span></code> columns are present.</p></li>
<li><p>Create <code class="docutils literal notranslate"><span class="pre">CHR</span></code> and/or <code class="docutils literal notranslate"><span class="pre">POS</span></code> column using a reference panel if <code class="docutils literal notranslate"><span class="pre">SNP</span></code> column is present.</p></li>
<li><p>Create <code class="docutils literal notranslate"><span class="pre">NEA</span></code> (non-effect allele) column using a reference panel if <code class="docutils literal notranslate"><span class="pre">EA</span></code> (effect allele) column is present.</p></li>
<li><p>Create the <code class="docutils literal notranslate"><span class="pre">SE</span></code> (standard-error) column if the <code class="docutils literal notranslate"><span class="pre">BETA</span></code> and <code class="docutils literal notranslate"><span class="pre">P</span></code> (p-value) columns are present.</p></li>
<li><p>Create the <code class="docutils literal notranslate"><span class="pre">P</span></code> column if the <code class="docutils literal notranslate"><span class="pre">BETA</span></code> and <code class="docutils literal notranslate"><span class="pre">SE</span></code> columns are present.</p></li>
</ul>
<p>If you do not wish to run certain steps, or wish to run only certain steps, you can use additional arguments. For more information, please refer to the <a class="reference internal" href="modules.html#id0" title="genal.Geno.preprocess_data"><code class="xref py py-meth docutils literal notranslate"><span class="pre">preprocess_data()</span></code></a> method in the API documentation.</p>
<p>In our case, the <code class="docutils literal notranslate"><span class="pre">SNP</span></code> column (for SNP identifier - rsid) was missing from our dataframe and has been added based on a 1000 genome reference panel:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">Using</span> <span class="n">the</span> <span class="n">EUR</span> <span class="n">reference</span> <span class="n">panel</span><span class="o">.</span>
<span class="n">The</span> <span class="n">SNP</span> <span class="n">column</span> <span class="p">(</span><span class="n">rsID</span><span class="p">)</span> <span class="n">has</span> <span class="n">been</span> <span class="n">created</span><span class="o">.</span> <span class="mi">197511</span> <span class="p">(</span><span class="mf">2.787</span><span class="o">%</span><span class="p">)</span> <span class="n">SNPs</span> <span class="n">were</span> <span class="ow">not</span> <span class="n">found</span> <span class="ow">in</span> <span class="n">the</span> <span class="n">reference</span> <span class="n">data</span> <span class="ow">and</span> <span class="n">their</span> <span class="n">ID</span> <span class="nb">set</span> <span class="n">to</span> <span class="n">CHR</span><span class="p">:</span><span class="n">POS</span><span class="p">:</span><span class="n">EA</span><span class="o">.</span>
<span class="n">The</span> <span class="n">BETA</span> <span class="n">column</span> <span class="n">looks</span> <span class="n">like</span> <span class="n">Beta</span> <span class="n">estimates</span><span class="o">.</span> <span class="n">Use</span> <span class="n">effect_column</span><span class="o">=</span><span class="s1">&#39;OR&#39;</span> <span class="k">if</span> <span class="n">it</span> <span class="ow">is</span> <span class="n">a</span> <span class="n">column</span> <span class="n">of</span> <span class="n">Odds</span> <span class="n">Ratios</span><span class="o">.</span>
</pre></div>
</div>
<p>You can always check the data of a <code class="docutils literal notranslate"><span class="pre">genal.Geno</span></code> instance by accessing the <code class="docutils literal notranslate"><span class="pre">data</span></code> attribute:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">SBP_Geno</span><span class="o">.</span><span class="n">data</span><span class="o">.</span><span class="n">head</span><span class="p">(</span><span class="mi">5</span><span class="p">)</span>
<span class="go">    EA NEA     EAF   BETA     SE        P  CHR       POS        SNP</span>
<span class="go">0    A   G  0.5660  0.0523  0.0303  0.083940   10  100000625  rs7899632</span>
<span class="go">1    A   C  0.7936  0.0200  0.0372  0.591100   10  100000645  rs61875309</span>
<span class="go">2    T   G  0.8831  0.1417  0.0469  0.002526   10  100003242  rs12258651</span>
<span class="go">3    A   G  0.9609  0.0245  0.0838  0.769800   10  100003304  rs72828461</span>
<span class="go">4    T   C  0.6406 -0.0680  0.0313  0.029870   10  100003785  rs1359508</span>
</pre></div>
</div>
<p>And we see that the <code class="docutils literal notranslate"><span class="pre">SNP</span></code> column with the rsids has been added based on the reference data. You do not need to obtain the 1000 genome reference panel yourself, genal will download it the first time you use it. By default, the reference panel used is the European (EUR) one. You can specify another valid reference panel (AFR, EAS, SAS, AMR) with the <code class="docutils literal notranslate"><span class="pre">reference_panel</span></code> argument:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">SBP_Geno</span><span class="o">.</span><span class="n">preprocess_data</span><span class="p">(</span><span class="n">preprocessing</span><span class="o">=</span><span class="s1">&#39;Fill_delete&#39;</span><span class="p">,</span> <span class="n">reference_panel</span><span class="o">=</span><span class="s2">&quot;afr&quot;</span><span class="p">)</span>
</pre></div>
</div>
<p>You can also use a custom reference panel by specifying the path to bed/bim/fam files (without the extension) in the <code class="docutils literal notranslate"><span class="pre">reference_panel</span></code> argument.</p>
</section>
</section>
<section id="clumping">
<h2>Clumping<a class="headerlink" href="#clumping" title="Link to this heading">¶</a></h2>
<p>Clumping, or C+T: Clumping + Thresholding, is the step at which we select the SNPs that will be used as our genetic instruments in future Polygenic Risk Scores and Mendelian Randomization analyses. The process involves identifying the SNPs that are strongly associated with our trait of interest (systolic blood pressure in this tutorial) and are independent from each other. This second step ensures that selected SNPs are not highly correlated, (i.e., they are not in high linkage disequilibrium). For this step, we again need to use a reference panel.</p>
<p>The SNP-data loaded in a <a class="reference internal" href="modules.html#genal.Geno" title="genal.Geno"><code class="xref py py-class docutils literal notranslate"><span class="pre">Geno</span></code></a> instance can be clumped using the <a class="reference internal" href="modules.html#id1" title="genal.Geno.clump"><code class="xref py py-meth docutils literal notranslate"><span class="pre">clump()</span></code></a> method. It will return another <a class="reference internal" href="modules.html#genal.Geno" title="genal.Geno"><code class="xref py py-class docutils literal notranslate"><span class="pre">Geno</span></code></a> instance containing only the clumped data:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">SBP_clumped</span> <span class="o">=</span> <span class="n">SBP_Geno</span><span class="o">.</span><span class="n">clump</span><span class="p">(</span><span class="n">p1</span><span class="o">=</span><span class="mf">5e-8</span><span class="p">,</span> <span class="n">r2</span><span class="o">=</span><span class="mf">0.1</span><span class="p">,</span> <span class="n">kb</span><span class="o">=</span><span class="mi">250</span><span class="p">,</span> <span class="n">reference_panel</span><span class="o">=</span><span class="s2">&quot;eur&quot;</span><span class="p">)</span>
</pre></div>
</div>
<p>It will output the number of instruments obtained:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">Using</span> <span class="n">the</span> <span class="n">EUR</span> <span class="n">reference</span> <span class="n">panel</span><span class="o">.</span>
<span class="ne">Warning</span><span class="p">:</span> <span class="mi">760</span>  <span class="n">top</span> <span class="n">variant</span> <span class="n">IDs</span> <span class="n">missing</span>
<span class="mi">1545</span> <span class="n">clumps</span> <span class="n">formed</span> <span class="kn">from</span> <span class="mi">73594</span> <span class="n">top</span> <span class="n">variants</span><span class="o">.</span>
</pre></div>
</div>
<p>You can specify the thresholds you want to use for the clumping with the following arguments:</p>
<ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">p1</span></code>: P-value threshold during clumping. SNPs with a P-value higher than this value are excluded. Defaults to <code class="docutils literal notranslate"><span class="pre">5e-8</span></code>.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">r2</span></code>: Linkage disequilibrium threshold for the independence check. Takes values between 0 and 1. Defaults to <code class="docutils literal notranslate"><span class="pre">0.1</span></code>.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">kb</span></code>: Genomic window used for the independence check (the unit is thousands of base-pair positions). Defaults to <code class="docutils literal notranslate"><span class="pre">250</span></code>.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">reference_panel</span></code>: The reference population used to derive linkage disequilibrium values and select independent SNPs. Defaults to <code class="docutils literal notranslate"><span class="pre">eur</span></code>.</p></li>
</ul>
</section>
<section id="polygenic-risk-scoring">
<h2>Polygenic Risk Scoring<a class="headerlink" href="#polygenic-risk-scoring" title="Link to this heading">¶</a></h2>
<p>Computing a Polygenic Risk Score (PRS) can be done in one line with the <a class="reference internal" href="modules.html#id2" title="genal.Geno.prs"><code class="xref py py-meth docutils literal notranslate"><span class="pre">prs()</span></code></a> method:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">SBP_clumped</span><span class="o">.</span><span class="n">prs</span><span class="p">(</span><span class="n">name</span><span class="o">=</span><span class="s2">&quot;SBP_prs&quot;</span><span class="p">,</span> <span class="n">path</span><span class="o">=</span><span class="s2">&quot;path/to/genetic/files&quot;</span><span class="p">)</span>
</pre></div>
</div>
<p>The genetic files of the target population can be either contained in one triple of bed/bim/fam files with information for all SNPs, or divided by chromosome (one bed/bim/fam triple for chr 1, another for chr 2, etc…). In the latter case, provide the path by replacing the chromosome number by <code class="docutils literal notranslate"><span class="pre">$</span></code> and genal will extract the necessary SNPs from each chromosome and merge them before running the PRS. For instance, if the genetic files are named <code class="docutils literal notranslate"><span class="pre">Pop_chr1.bed</span></code>, <code class="docutils literal notranslate"><span class="pre">Pop_chr1.bim</span></code>, <code class="docutils literal notranslate"><span class="pre">Pop_chr1.fam</span></code>, <code class="docutils literal notranslate"><span class="pre">Pop_chr2.bed</span></code>, …, you can use:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">SBP_clumped</span><span class="o">.</span><span class="n">prs</span><span class="p">(</span><span class="n">name</span><span class="o">=</span><span class="s2">&quot;SBP_prs&quot;</span><span class="p">,</span> <span class="n">path</span><span class="o">=</span><span class="s2">&quot;Pop_chr$&quot;</span><span class="p">)</span>
</pre></div>
</div>
<p>The <code class="docutils literal notranslate"><span class="pre">name</span></code> argument specifies the name of the .csv file that will be saved with the individual risk scores.
The output of the <a class="reference internal" href="modules.html#id2" title="genal.Geno.prs"><code class="xref py py-meth docutils literal notranslate"><span class="pre">prs()</span></code></a> method will include how many SNPs were used to compute the risk score. It can happen that some of the SNPs are multiallelic in the genetic data (even if they are not multiallelic in our SNP data) and need to be excluded. It can also happen that some of the SNPs are missing from the genetic files of the target population (for instance if the data has not been imputed):</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">CHR</span><span class="o">/</span><span class="n">POS</span> <span class="n">columns</span> <span class="n">present</span><span class="p">:</span> <span class="n">SNPs</span> <span class="n">searched</span> <span class="n">based</span> <span class="n">on</span> <span class="n">genomic</span> <span class="n">positions</span><span class="o">.</span>
<span class="n">Extracting</span> <span class="n">SNPs</span> <span class="k">for</span> <span class="n">each</span> <span class="n">chromosome</span><span class="o">...</span>
<span class="n">SNPs</span> <span class="n">extracted</span> <span class="k">for</span> <span class="n">chr1</span><span class="o">.</span>
<span class="n">SNPs</span> <span class="n">extracted</span> <span class="k">for</span> <span class="n">chr2</span><span class="o">.</span>
<span class="n">SNPs</span> <span class="n">extracted</span> <span class="k">for</span> <span class="n">chr3</span><span class="o">.</span>
<span class="n">SNPs</span> <span class="n">extracted</span> <span class="k">for</span> <span class="n">chr4</span><span class="o">.</span>
<span class="n">SNPs</span> <span class="n">extracted</span> <span class="k">for</span> <span class="n">chr5</span><span class="o">.</span>
<span class="n">SNPs</span> <span class="n">extracted</span> <span class="k">for</span> <span class="n">chr6</span><span class="o">.</span>
<span class="n">SNPs</span> <span class="n">extracted</span> <span class="k">for</span> <span class="n">chr7</span><span class="o">.</span>
<span class="n">SNPs</span> <span class="n">extracted</span> <span class="k">for</span> <span class="n">chr8</span><span class="o">.</span>
<span class="n">SNPs</span> <span class="n">extracted</span> <span class="k">for</span> <span class="n">chr9</span><span class="o">.</span>
<span class="n">SNPs</span> <span class="n">extracted</span> <span class="k">for</span> <span class="n">chr10</span><span class="o">.</span>
<span class="n">SNPs</span> <span class="n">extracted</span> <span class="k">for</span> <span class="n">chr11</span><span class="o">.</span>
<span class="n">SNPs</span> <span class="n">extracted</span> <span class="k">for</span> <span class="n">chr12</span><span class="o">.</span>
<span class="n">SNPs</span> <span class="n">extracted</span> <span class="k">for</span> <span class="n">chr13</span><span class="o">.</span>
<span class="n">SNPs</span> <span class="n">extracted</span> <span class="k">for</span> <span class="n">chr14</span><span class="o">.</span>
<span class="n">SNPs</span> <span class="n">extracted</span> <span class="k">for</span> <span class="n">chr15</span><span class="o">.</span>
<span class="n">SNPs</span> <span class="n">extracted</span> <span class="k">for</span> <span class="n">chr16</span><span class="o">.</span>
<span class="n">SNPs</span> <span class="n">extracted</span> <span class="k">for</span> <span class="n">chr17</span><span class="o">.</span>
<span class="n">SNPs</span> <span class="n">extracted</span> <span class="k">for</span> <span class="n">chr18</span><span class="o">.</span>
<span class="n">SNPs</span> <span class="n">extracted</span> <span class="k">for</span> <span class="n">chr19</span><span class="o">.</span>
<span class="n">SNPs</span> <span class="n">extracted</span> <span class="k">for</span> <span class="n">chr20</span><span class="o">.</span>
<span class="n">SNPs</span> <span class="n">extracted</span> <span class="k">for</span> <span class="n">chr21</span><span class="o">.</span>
<span class="n">SNPs</span> <span class="n">extracted</span> <span class="k">for</span> <span class="n">chr22</span><span class="o">.</span>
<span class="n">Merging</span> <span class="n">SNPs</span> <span class="n">extracted</span> <span class="kn">from</span> <span class="nn">each</span> <span class="n">chromosome</span><span class="o">...</span>
<span class="n">Created</span> <span class="n">bed</span><span class="o">/</span><span class="n">bim</span><span class="o">/</span><span class="n">fam</span> <span class="n">fileset</span> <span class="k">with</span> <span class="n">extracted</span> <span class="n">SNPs</span><span class="p">:</span> <span class="n">tmp_GENAL</span><span class="o">/</span><span class="mi">4</span><span class="n">f4ce6a7_allchr</span>
<span class="n">Extraction</span> <span class="n">completed</span><span class="o">.</span> <span class="mi">786</span><span class="p">(</span><span class="mf">50.874</span><span class="o">%</span><span class="p">)</span> <span class="n">SNPs</span> <span class="n">were</span> <span class="ow">not</span> <span class="n">extracted</span> <span class="kn">from</span> <span class="nn">the</span> <span class="n">genetic</span> <span class="n">data</span><span class="o">.</span>
<span class="n">Computing</span> <span class="n">a</span> <span class="n">weighted</span> <span class="n">PRS</span> <span class="n">using</span> <span class="n">tmp_GENAL</span><span class="o">/</span><span class="mi">4</span><span class="n">f4ce6a7_allchr</span> <span class="n">data</span><span class="o">.</span>
<span class="n">The</span> <span class="n">PRS</span> <span class="n">computation</span> <span class="n">was</span> <span class="n">successful</span> <span class="ow">and</span> <span class="n">used</span> <span class="mi">759</span><span class="o">/</span><span class="mi">1545</span> <span class="p">(</span><span class="mf">49.126</span><span class="o">%</span><span class="p">)</span> <span class="n">SNPs</span><span class="o">.</span>
<span class="n">PRS</span> <span class="n">data</span> <span class="n">saved</span> <span class="n">to</span> <span class="n">SBP_prs</span><span class="o">.</span><span class="n">csv</span>
</pre></div>
</div>
<p>Here, we see that about half of the SNPs were not extracted from the data. In such cases, we may want to try and salvage some of these SNPs by looking for proxies (SNPs in high linkage disequilibrium, i.e. highly correlated SNPs). This can be done by specifying the <code class="docutils literal notranslate"><span class="pre">proxy</span> <span class="pre">=</span> <span class="pre">True</span></code> argument:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">SBP_clumped</span><span class="o">.</span><span class="n">prs</span><span class="p">(</span><span class="n">name</span><span class="o">=</span><span class="s2">&quot;SBP_prs_proxy&quot;</span><span class="p">,</span> <span class="n">path</span><span class="o">=</span><span class="s2">&quot;Pop_chr$&quot;</span><span class="p">,</span> <span class="n">proxy</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">reference_panel</span><span class="o">=</span><span class="s2">&quot;eur&quot;</span><span class="p">,</span> <span class="n">r2</span><span class="o">=</span><span class="mf">0.8</span><span class="p">,</span> <span class="n">kb</span><span class="o">=</span><span class="mi">5000</span><span class="p">,</span> <span class="n">window_snps</span><span class="o">=</span><span class="mi">5000</span><span class="p">)</span>
</pre></div>
</div>
<p>and the output is:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">CHR</span><span class="o">/</span><span class="n">POS</span> <span class="n">columns</span> <span class="n">present</span><span class="p">:</span> <span class="n">SNPs</span> <span class="n">searched</span> <span class="n">based</span> <span class="n">on</span> <span class="n">genomic</span> <span class="n">positions</span><span class="o">.</span>
<span class="n">Identifying</span> <span class="n">the</span> <span class="n">SNPs</span> <span class="n">present</span> <span class="ow">in</span> <span class="n">the</span> <span class="n">genetic</span> <span class="n">data</span><span class="o">...</span>
<span class="mi">759</span> <span class="n">SNPs</span> <span class="n">out</span> <span class="n">of</span> <span class="mi">1545</span> <span class="n">are</span> <span class="n">present</span> <span class="ow">in</span> <span class="n">the</span> <span class="n">genetic</span> <span class="n">data</span><span class="o">.</span>
<span class="n">Searching</span> <span class="n">proxies</span> <span class="k">for</span> <span class="mi">786</span> <span class="n">SNPs</span><span class="o">...</span>
<span class="n">Using</span> <span class="n">the</span> <span class="n">EUR</span> <span class="n">reference</span> <span class="n">panel</span><span class="o">.</span>
<span class="n">Filtering</span> <span class="n">the</span> <span class="n">potential</span> <span class="n">proxies</span> <span class="k">with</span> <span class="n">the</span> <span class="n">searchspace</span> <span class="n">provided</span><span class="o">.</span>
<span class="n">Found</span> <span class="n">proxies</span> <span class="k">for</span> <span class="mi">578</span> <span class="n">missing</span> <span class="n">SNPs</span><span class="o">.</span>
<span class="mi">7</span><span class="p">(</span><span class="mf">0.455</span><span class="o">%</span><span class="p">)</span> <span class="n">duplicated</span> <span class="n">SNPs</span> <span class="n">have</span> <span class="n">been</span> <span class="n">removed</span><span class="o">.</span> <span class="n">Use</span> <span class="n">keep_dups</span><span class="o">=</span><span class="kc">True</span> <span class="n">to</span> <span class="n">keep</span> <span class="n">them</span><span class="o">.</span>
<span class="n">Extracting</span> <span class="n">SNPs</span> <span class="k">for</span> <span class="n">each</span> <span class="n">chromosome</span><span class="o">...</span>
<span class="n">SNPs</span> <span class="n">extracted</span> <span class="k">for</span> <span class="n">chr1</span><span class="o">.</span>
<span class="n">SNPs</span> <span class="n">extracted</span> <span class="k">for</span> <span class="n">chr2</span><span class="o">.</span>
<span class="n">SNPs</span> <span class="n">extracted</span> <span class="k">for</span> <span class="n">chr3</span><span class="o">.</span>
<span class="n">SNPs</span> <span class="n">extracted</span> <span class="k">for</span> <span class="n">chr4</span><span class="o">.</span>
<span class="n">SNPs</span> <span class="n">extracted</span> <span class="k">for</span> <span class="n">chr5</span><span class="o">.</span>
<span class="n">SNPs</span> <span class="n">extracted</span> <span class="k">for</span> <span class="n">chr6</span><span class="o">.</span>
<span class="n">SNPs</span> <span class="n">extracted</span> <span class="k">for</span> <span class="n">chr7</span><span class="o">.</span>
<span class="n">SNPs</span> <span class="n">extracted</span> <span class="k">for</span> <span class="n">chr8</span><span class="o">.</span>
<span class="n">SNPs</span> <span class="n">extracted</span> <span class="k">for</span> <span class="n">chr9</span><span class="o">.</span>
<span class="n">SNPs</span> <span class="n">extracted</span> <span class="k">for</span> <span class="n">chr10</span><span class="o">.</span>
<span class="n">SNPs</span> <span class="n">extracted</span> <span class="k">for</span> <span class="n">chr11</span><span class="o">.</span>
<span class="n">SNPs</span> <span class="n">extracted</span> <span class="k">for</span> <span class="n">chr12</span><span class="o">.</span>
<span class="n">SNPs</span> <span class="n">extracted</span> <span class="k">for</span> <span class="n">chr13</span><span class="o">.</span>
<span class="n">SNPs</span> <span class="n">extracted</span> <span class="k">for</span> <span class="n">chr14</span><span class="o">.</span>
<span class="n">SNPs</span> <span class="n">extracted</span> <span class="k">for</span> <span class="n">chr15</span><span class="o">.</span>
<span class="n">SNPs</span> <span class="n">extracted</span> <span class="k">for</span> <span class="n">chr16</span><span class="o">.</span>
<span class="n">SNPs</span> <span class="n">extracted</span> <span class="k">for</span> <span class="n">chr17</span><span class="o">.</span>
<span class="n">SNPs</span> <span class="n">extracted</span> <span class="k">for</span> <span class="n">chr18</span><span class="o">.</span>
<span class="n">SNPs</span> <span class="n">extracted</span> <span class="k">for</span> <span class="n">chr19</span><span class="o">.</span>
<span class="n">SNPs</span> <span class="n">extracted</span> <span class="k">for</span> <span class="n">chr20</span><span class="o">.</span>
<span class="n">SNPs</span> <span class="n">extracted</span> <span class="k">for</span> <span class="n">chr21</span><span class="o">.</span>
<span class="n">SNPs</span> <span class="n">extracted</span> <span class="k">for</span> <span class="n">chr22</span><span class="o">.</span>
<span class="n">Merging</span> <span class="n">SNPs</span> <span class="n">extracted</span> <span class="kn">from</span> <span class="nn">each</span> <span class="n">chromosome</span><span class="o">...</span>
<span class="n">Created</span> <span class="n">bed</span><span class="o">/</span><span class="n">bim</span><span class="o">/</span><span class="n">fam</span> <span class="n">fileset</span> <span class="k">with</span> <span class="n">extracted</span> <span class="n">SNPs</span><span class="p">:</span> <span class="n">tmp_GENAL</span><span class="o">/</span><span class="mi">4</span><span class="n">f4ce6a7_allchr</span>
<span class="n">Extraction</span> <span class="n">completed</span><span class="o">.</span> <span class="mi">208</span><span class="p">(</span><span class="mf">13.524</span><span class="o">%</span><span class="p">)</span> <span class="n">SNPs</span> <span class="n">were</span> <span class="ow">not</span> <span class="n">extracted</span> <span class="kn">from</span> <span class="nn">the</span> <span class="n">genetic</span> <span class="n">data</span><span class="o">.</span>
<span class="n">Computing</span> <span class="n">a</span> <span class="n">weighted</span> <span class="n">PRS</span> <span class="n">using</span> <span class="n">tmp_GENAL</span><span class="o">/</span><span class="mi">4</span><span class="n">f4ce6a7_allchr</span> <span class="n">data</span><span class="o">.</span>
<span class="n">The</span> <span class="n">PRS</span> <span class="n">computation</span> <span class="n">was</span> <span class="n">successful</span> <span class="ow">and</span> <span class="n">used</span> <span class="mi">1330</span><span class="o">/</span><span class="mi">1538</span> <span class="p">(</span><span class="mf">86.476</span><span class="o">%</span><span class="p">)</span> <span class="n">SNPs</span><span class="o">.</span>
<span class="n">PRS</span> <span class="n">data</span> <span class="n">saved</span> <span class="n">to</span> <span class="n">SBP_prs</span><span class="o">.</span><span class="n">csv</span>
</pre></div>
</div>
<p>In our case, we have been able to find proxies for 578 of the 786 SNPs that were missing in the population genetic data (7 potential proxies have been removed because they were identical to SNPs already present in our data).</p>
<p>You can customize how the proxies are chosen with the following arguments:</p>
<ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">reference_panel</span></code>: The reference population used to derive linkage disequilibrium values and find proxies. Defaults to <code class="docutils literal notranslate"><span class="pre">eur</span></code>.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">kb</span></code>: Width of the genomic window to look for proxies (in thousands of base-pair positions). Defaults to <code class="docutils literal notranslate"><span class="pre">5000</span></code>.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">r2</span></code>: Minimum linkage disequilibrium value with the original SNP for a proxy to be included. Defaults to <code class="docutils literal notranslate"><span class="pre">0.8</span></code>.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">window_snps</span></code>: Width of the window to look for proxies (in number of SNPs). Defaults to <code class="docutils literal notranslate"><span class="pre">5000</span></code>.</p></li>
</ul>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>You can call the <a class="reference internal" href="modules.html#id2" title="genal.Geno.prs"><code class="xref py py-meth docutils literal notranslate"><span class="pre">prs()</span></code></a> method on any <a class="reference internal" href="modules.html#genal.Geno" title="genal.Geno"><code class="xref py py-class docutils literal notranslate"><span class="pre">Geno</span></code></a> instance (containing at least the EA, BETA, and either SNP or CHR/POS columns). The data does not need to be clumped, and there is no limit to the number of SNPs used to compute the scores.</p>
</div>
</section>
<section id="mendelian-randomization">
<h2>Mendelian Randomization<a class="headerlink" href="#mendelian-randomization" title="Link to this heading">¶</a></h2>
<p>To run MR, we need to load both our exposure and outcome SNP-level data in <a class="reference internal" href="modules.html#genal.Geno" title="genal.Geno"><code class="xref py py-class docutils literal notranslate"><span class="pre">Geno</span></code></a> instances. In our case, the genetic instruments of the MR are the SNPs associated with blood pressure at genome-wide significant levels resulting from the clumping of the blood pressure GWAS. They are stored in our <code class="docutils literal notranslate"><span class="pre">SBP_clumped</span></code> <a class="reference internal" href="modules.html#genal.Geno" title="genal.Geno"><code class="xref py py-class docutils literal notranslate"><span class="pre">Geno</span></code></a> instance which also include their association with the exposure trait (instrument-SBP estimates in the <code class="docutils literal notranslate"><span class="pre">BETA</span></code> column).</p>
<p>To get their association with the outcome trait (instrument-stroke estimates), we are going to use SNP-level data from a large GWAS of stroke performed by the GIGASTROKE consortium (<a class="reference external" href="https://www.nature.com/articles/s41586-022-05165-3">Nature article</a>):</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">stroke_gwas</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="s2">&quot;GCST90104539_buildGRCh37.tsv&quot;</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">)</span>
</pre></div>
</div>
<p>We inspect it to determine the column names:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">chromosome</span>  <span class="n">base_pair_location</span>  <span class="n">effect_allele_frequency</span>   <span class="n">beta</span>  <span class="n">standard_error</span>  <span class="n">p_value</span>  <span class="n">odds_ratio</span>  <span class="n">ci_lower</span>  <span class="n">ci_upper</span> <span class="n">effect_allele</span> <span class="n">other_allele</span>
<span class="mi">0</span>           <span class="mi">5</span>            <span class="mi">29439275</span>                    <span class="mf">0.3569</span>  <span class="mf">0.0030</span>         <span class="mf">0.0070</span>  <span class="mf">0.6658</span>   <span class="mf">1.003005</span>  <span class="mf">0.989337</span>  <span class="mf">1.016861</span>            <span class="n">T</span>            <span class="n">C</span>
<span class="mi">1</span>           <span class="mi">5</span>            <span class="mi">85928892</span>                    <span class="mf">0.0639</span> <span class="o">-</span><span class="mf">0.0152</span>         <span class="mf">0.0137</span>  <span class="mf">0.2686</span>   <span class="mf">0.984915</span>  <span class="mf">0.958820</span>  <span class="mf">1.011720</span>            <span class="n">T</span>            <span class="n">C</span>
<span class="mi">2</span>          <span class="mi">10</span>           <span class="mi">128341232</span>                    <span class="mf">0.4613</span>  <span class="mf">0.0025</span>         <span class="mf">0.0065</span>  <span class="mf">0.6998</span>   <span class="mf">1.002503</span>  <span class="mf">0.989812</span>  <span class="mf">1.015357</span>            <span class="n">T</span>            <span class="n">C</span>
<span class="mi">3</span>           <span class="mi">3</span>            <span class="mi">62707519</span>                    <span class="mf">0.0536</span>  <span class="mf">0.0152</span>         <span class="mf">0.0152</span>  <span class="mf">0.3177</span>   <span class="mf">1.015316</span>  <span class="mf">0.985514</span>  <span class="mf">1.046019</span>            <span class="n">T</span>            <span class="n">C</span>
<span class="mi">4</span>           <span class="mi">2</span>            <span class="mi">80464120</span>                    <span class="mf">0.9789</span>  <span class="mf">0.0057</span>         <span class="mf">0.0254</span>  <span class="mf">0.8223</span>   <span class="mf">1.005716</span>  <span class="mf">0.956874</span>  <span class="mf">1.057052</span>            <span class="n">T</span>            <span class="n">G</span>
</pre></div>
</div>
<p>We load it in a <a class="reference internal" href="modules.html#genal.Geno" title="genal.Geno"><code class="xref py py-class docutils literal notranslate"><span class="pre">Geno</span></code></a> instance:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">Stroke_Geno</span> <span class="o">=</span> <span class="n">genal</span><span class="o">.</span><span class="n">Geno</span><span class="p">(</span><span class="n">stroke_gwas</span><span class="p">,</span> <span class="n">CHR</span><span class="o">=</span><span class="s2">&quot;chromosome&quot;</span><span class="p">,</span> <span class="n">POS</span><span class="o">=</span><span class="s2">&quot;base_pair_location&quot;</span><span class="p">,</span>
                         <span class="n">EA</span><span class="o">=</span><span class="s2">&quot;effect_allele&quot;</span><span class="p">,</span> <span class="n">NEA</span><span class="o">=</span><span class="s2">&quot;other_allele&quot;</span><span class="p">,</span> <span class="n">BETA</span><span class="o">=</span><span class="s2">&quot;beta&quot;</span><span class="p">,</span>
                         <span class="n">SE</span><span class="o">=</span><span class="s2">&quot;standard_error&quot;</span><span class="p">,</span> <span class="n">P</span><span class="o">=</span><span class="s2">&quot;p_value&quot;</span><span class="p">,</span>
                         <span class="n">EAF</span><span class="o">=</span><span class="s2">&quot;effect_allele_frequency&quot;</span><span class="p">,</span> <span class="n">keep_columns</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
</pre></div>
</div>
<p>We preprocess it as well to put it in the correct format and make sure there are no invalid values:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">Stroke_Geno</span><span class="o">.</span><span class="n">preprocess_data</span><span class="p">(</span><span class="n">preprocessing</span><span class="o">=</span><span class="s1">&#39;Fill_delete&#39;</span><span class="p">)</span>
</pre></div>
</div>
<p>Now, we need to extract our instruments (SNPs of the <code class="docutils literal notranslate"><span class="pre">SBP_clumped</span></code> data) from the outcome data to obtain their association with the outcome trait (stroke). It can be done by calling the <a class="reference internal" href="modules.html#id3" title="genal.Geno.query_outcome"><code class="xref py py-meth docutils literal notranslate"><span class="pre">query_outcome()</span></code></a> method:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">SBP_clumped</span><span class="o">.</span><span class="n">query_outcome</span><span class="p">(</span><span class="n">Stroke_Geno</span><span class="p">,</span> <span class="n">proxy</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
</pre></div>
</div>
<p>Genal will print how many SNPs were successfully found and extracted from the outcome data:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">Outcome</span> <span class="n">data</span> <span class="n">successfully</span> <span class="n">loaded</span> <span class="kn">from</span> <span class="s1">&#39;b352e412&#39;</span> <span class="n">geno</span> <span class="n">instance</span><span class="o">.</span>
<span class="n">Identifying</span> <span class="n">the</span> <span class="n">exposure</span> <span class="n">SNPs</span> <span class="n">present</span> <span class="ow">in</span> <span class="n">the</span> <span class="n">outcome</span> <span class="n">data</span><span class="o">...</span>
<span class="mi">1541</span> <span class="n">SNPs</span> <span class="n">out</span> <span class="n">of</span> <span class="mi">1545</span> <span class="n">are</span> <span class="n">present</span> <span class="ow">in</span> <span class="n">the</span> <span class="n">outcome</span> <span class="n">data</span><span class="o">.</span>
<span class="p">(</span><span class="n">Exposure</span> <span class="n">data</span><span class="p">,</span> <span class="n">Outcome</span> <span class="n">data</span><span class="p">,</span> <span class="n">Outcome</span> <span class="n">name</span><span class="p">)</span> <span class="n">stored</span> <span class="ow">in</span> <span class="n">the</span> <span class="o">.</span><span class="n">MR_data</span> <span class="n">attribute</span><span class="o">.</span>
</pre></div>
</div>
<p>Here as well you have the option to use proxies for the instruments that are not present in the outcome data:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">SBP_clumped</span><span class="o">.</span><span class="n">query_outcome</span><span class="p">(</span><span class="n">Stroke_geno</span><span class="p">,</span> <span class="n">proxy</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">reference_panel</span><span class="o">=</span><span class="s2">&quot;eur&quot;</span><span class="p">,</span>
                          <span class="n">kb</span><span class="o">=</span><span class="mi">5000</span><span class="p">,</span> <span class="n">r2</span><span class="o">=</span><span class="mf">0.6</span><span class="p">,</span> <span class="n">window_snps</span><span class="o">=</span><span class="mi">5000</span><span class="p">)</span>
</pre></div>
</div>
<p>And genal will print the number of missing instruments which have been proxied:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">Outcome</span> <span class="n">data</span> <span class="n">successfully</span> <span class="n">loaded</span> <span class="kn">from</span> <span class="s1">&#39;b352e412&#39;</span> <span class="n">geno</span> <span class="n">instance</span><span class="o">.</span>
<span class="n">Identifying</span> <span class="n">the</span> <span class="n">exposure</span> <span class="n">SNPs</span> <span class="n">present</span> <span class="ow">in</span> <span class="n">the</span> <span class="n">outcome</span> <span class="n">data</span><span class="o">...</span>
<span class="mi">1541</span> <span class="n">SNPs</span> <span class="n">out</span> <span class="n">of</span> <span class="mi">1545</span> <span class="n">are</span> <span class="n">present</span> <span class="ow">in</span> <span class="n">the</span> <span class="n">outcome</span> <span class="n">data</span><span class="o">.</span>
<span class="n">Searching</span> <span class="n">proxies</span> <span class="k">for</span> <span class="mi">4</span> <span class="n">SNPs</span><span class="o">...</span>
<span class="n">Using</span> <span class="n">the</span> <span class="n">EUR</span> <span class="n">reference</span> <span class="n">panel</span><span class="o">.</span>
<span class="n">Found</span> <span class="n">proxies</span> <span class="k">for</span> <span class="mi">4</span> <span class="n">SNPs</span><span class="o">.</span>
<span class="p">(</span><span class="n">Exposure</span> <span class="n">data</span><span class="p">,</span> <span class="n">Outcome</span> <span class="n">data</span><span class="p">,</span> <span class="n">Outcome</span> <span class="n">name</span><span class="p">)</span> <span class="n">stored</span> <span class="ow">in</span> <span class="n">the</span> <span class="o">.</span><span class="n">MR_data</span> <span class="n">attribute</span><span class="o">.</span>
</pre></div>
</div>
<p>After extracting the instruments from the outcome data, the <code class="docutils literal notranslate"><span class="pre">SBP_clumped</span></code> <a class="reference internal" href="modules.html#genal.Geno" title="genal.Geno"><code class="xref py py-class docutils literal notranslate"><span class="pre">Geno</span></code></a> instance contains an <a class="reference internal" href="modules.html#id4" title="genal.Geno.MR"><code class="xref py py-attr docutils literal notranslate"><span class="pre">MR</span></code></a> attribute containing the instruments-exposure and instruments-outcome associations necessary to run MR. Running MR is now as simple as calling the <a class="reference internal" href="modules.html#id4" title="genal.Geno.MR"><code class="xref py py-meth docutils literal notranslate"><span class="pre">MR()</span></code></a> method of the SBP_clumped <a class="reference internal" href="modules.html#genal.Geno" title="genal.Geno"><code class="xref py py-class docutils literal notranslate"><span class="pre">Geno</span></code></a> instance:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">SBP_clumped</span><span class="o">.</span><span class="n">MR</span><span class="p">(</span><span class="n">action</span><span class="o">=</span><span class="mi">2</span><span class="p">,</span> <span class="n">exposure_name</span><span class="o">=</span><span class="s2">&quot;SBP&quot;</span><span class="p">,</span> <span class="n">outcome_name</span><span class="o">=</span><span class="s2">&quot;Stroke_eur&quot;</span><span class="p">)</span>
</pre></div>
</div>
<p>The <a class="reference internal" href="modules.html#id4" title="genal.Geno.MR"><code class="xref py py-meth docutils literal notranslate"><span class="pre">MR()</span></code></a> method prints a message specifying which SNPs have been excluded from the analysis (it depends on the action argument, as we will see):</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">Action</span> <span class="o">=</span> <span class="mi">2</span><span class="p">:</span> <span class="mi">42</span> <span class="n">SNPs</span> <span class="n">excluded</span> <span class="k">for</span> <span class="n">being</span> <span class="n">palindromic</span> <span class="k">with</span> <span class="n">intermediate</span> <span class="n">allele</span> <span class="n">frequencies</span><span class="p">:</span> <span class="n">rs11817866</span><span class="p">,</span> <span class="n">rs3802517</span><span class="p">,</span> <span class="n">rs2788293</span><span class="p">,</span> <span class="n">rs2274224</span><span class="p">,</span> <span class="n">rs7310615</span><span class="p">,</span> <span class="n">rs7953257</span><span class="p">,</span> <span class="n">rs2024385</span><span class="p">,</span> <span class="n">rs61912333</span><span class="p">,</span> <span class="n">rs11632436</span><span class="p">,</span> <span class="n">rs1012089</span><span class="p">,</span> <span class="n">rs3851018</span><span class="p">,</span> <span class="n">rs9899540</span><span class="p">,</span> <span class="n">rs4617956</span><span class="p">,</span> <span class="n">rs773432</span><span class="p">,</span> <span class="n">rs11585169</span><span class="p">,</span> <span class="n">rs7796</span><span class="p">,</span> <span class="n">rs2487904</span><span class="p">,</span> <span class="n">rs12321</span><span class="p">,</span> <span class="n">rs73029563</span><span class="p">,</span> <span class="n">rs4673238</span><span class="p">,</span> <span class="n">rs3845811</span><span class="p">,</span> <span class="n">rs2160236</span><span class="p">,</span> <span class="n">rs10165271</span><span class="p">,</span> <span class="n">rs9848170</span><span class="p">,</span> <span class="n">rs2724535</span><span class="p">,</span> <span class="n">rs6842486</span><span class="p">,</span> <span class="n">rs4834792</span><span class="p">,</span> <span class="n">rs990619</span><span class="p">,</span> <span class="n">rs155364</span><span class="p">,</span> <span class="n">rs480882</span><span class="p">,</span> <span class="n">rs6875372</span><span class="p">,</span> <span class="n">rs258951</span><span class="p">,</span> <span class="n">rs1870735</span><span class="p">,</span> <span class="n">rs1800795</span><span class="p">,</span> <span class="n">rs12700814</span><span class="p">,</span> <span class="n">rs1821002</span><span class="p">,</span> <span class="n">rs3021500</span><span class="p">,</span> <span class="n">rs28601761</span><span class="p">,</span> <span class="n">rs7463212</span><span class="p">,</span> <span class="n">rs907183</span><span class="p">,</span> <span class="n">rs534523</span><span class="p">,</span> <span class="n">rs520015</span>
</pre></div>
</div>
<p>It returns a dataframe containing the results for different MR methods:</p>
<table class="docutils align-default">
<thead>
<tr class="row-odd"><th class="head"><p>exposure</p></th>
<th class="head"><p>outcome</p></th>
<th class="head"><p>method</p></th>
<th class="head"><p>nSNP</p></th>
<th class="head"><p>b</p></th>
<th class="head"><p>se</p></th>
<th class="head"><p>pval</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>SBP</p></td>
<td><p>Stroke_eur</p></td>
<td><p>Inverse-Variance Weighted</p></td>
<td><p>1499</p></td>
<td><p>0.023049</p></td>
<td><p>0.001061</p></td>
<td><p>1.382645e-104</p></td>
</tr>
<tr class="row-odd"><td><p>SBP</p></td>
<td><p>Stroke_eur</p></td>
<td><p>Inverse Variance Weighted (Fixed Effects)</p></td>
<td><p>1499</p></td>
<td><p>0.023049</p></td>
<td><p>0.000754</p></td>
<td><p>4.390655e-205</p></td>
</tr>
<tr class="row-even"><td><p>SBP</p></td>
<td><p>Stroke_eur</p></td>
<td><p>Weighted Median</p></td>
<td><p>1499</p></td>
<td><p>0.022365</p></td>
<td><p>0.001337</p></td>
<td><p>8.863203e-63</p></td>
</tr>
<tr class="row-odd"><td><p>SBP</p></td>
<td><p>Stroke_eur</p></td>
<td><p>Simple mode</p></td>
<td><p>1499</p></td>
<td><p>0.027125</p></td>
<td><p>0.007698</p></td>
<td><p>4.382993e-04</p></td>
</tr>
<tr class="row-even"><td><p>SBP</p></td>
<td><p>Stroke_eur</p></td>
<td><p>MR Egger</p></td>
<td><p>1499</p></td>
<td><p>0.027543</p></td>
<td><p>0.002849</p></td>
<td><p>1.723156e-21</p></td>
</tr>
<tr class="row-odd"><td><p>SBP</p></td>
<td><p>Stroke_eur</p></td>
<td><p>Egger Intercept</p></td>
<td><p>1499</p></td>
<td><p>-0.001381</p></td>
<td><p>0.000813</p></td>
<td><p>8.935529e-02</p></td>
</tr>
</tbody>
</table>
<p>You can specify several arguments. We refer to the API for a full list, but the most important one is the <code class="docutils literal notranslate"><span class="pre">action</span></code> argument. It determines how palindromic SNPs are treated during the exposure-outcome harmonization step. Palindromic SNPs are SNPs where the nucleotide change reads the same forward and backward on complementary strands of DNA (for instance <code class="docutils literal notranslate"><span class="pre">EA</span> <span class="pre">=</span> <span class="pre">'A'</span></code> and <code class="docutils literal notranslate"><span class="pre">NEA</span> <span class="pre">=</span> <span class="pre">'T'</span></code>).</p>
<ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">action</span> <span class="pre">=</span> <span class="pre">1</span></code>: Palindromic SNPs are not treated (assumes all alleles are on the forward strand)</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">action</span> <span class="pre">=</span> <span class="pre">2</span></code>: Uses effect allele frequencies to attempt to flip them (conservative, default)</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">action</span> <span class="pre">=</span> <span class="pre">3</span></code>: Removes all palindromic SNPs (very conservative)</p></li>
</ul>
<p>If you choose the option 2 or 3 (recommended), genal will print the list of palindromic SNPs that have been removed from the analysis.</p>
<p>By default, only some MR methods (inverse-variance weighted, weighted median, Simple mode, MR-Egger) are going to be run. But if you wish to run a different set of MR methods, you can pass a list of strings to the <code class="docutils literal notranslate"><span class="pre">methods</span></code> argument. The possible strings are:</p>
<ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">IVW</span></code> for the classical Inverse-Variance Weighted method with random effects</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">IVW-RE</span></code> for the Inverse Variance Weighted method with Random Effects where the standard error is not corrected for under dispersion</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">IVW-FE</span></code> for the Inverse Variance Weighted with fixed effects</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">UWR</span></code> for the Unweighted Regression method</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">WM</span></code> for the Weighted Median method</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">WM-pen</span></code> for the penalised Weighted Median method</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">Simple-median</span></code> for the Simple Median method</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">Sign</span></code> for the Sign concordance test</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">Egger</span></code> for MR-Egger and the MR-Egger intercept</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">Egger-boot</span></code> for the bootstrapped version of MR-Egger and its intercept</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">Simple-mode</span></code> for the Simple mode method</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">Weighted-mode</span></code> for the Weighted mode method</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">all</span></code> to run all the above methods</p></li>
</ul>
<p>For more fine-tuning, such as settings for the number of boostrapping iterations, please refer to the API.</p>
<p>If you want to visualize the obtained MR results, you can use the <a class="reference internal" href="modules.html#genal.Geno.MR_plot" title="genal.Geno.MR_plot"><code class="xref py py-meth docutils literal notranslate"><span class="pre">MR_plot()</span></code></a> method that will plot each SNP in an <code class="docutils literal notranslate"><span class="pre">effect_on_exposure</span> <span class="pre">x</span> <span class="pre">effect_on_outcome</span></code> plane as well as lines corresponding to different MR methods:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">SBP_clumped</span><span class="o">.</span><span class="n">MR_plot</span><span class="p">(</span><span class="n">filename</span><span class="o">=</span><span class="s2">&quot;MR_plot_SBP_AS&quot;</span><span class="p">)</span>
</pre></div>
</div>
<img alt="MR plot" src="_images/MR_plot_SBP_AS.png" />
<p>You can select which MR methods you wish to plot with the <code class="docutils literal notranslate"><span class="pre">methods</span></code> argument. Note that for an MR method to be plotted, they must be included in the latest <a class="reference internal" href="modules.html#id4" title="genal.Geno.MR"><code class="xref py py-meth docutils literal notranslate"><span class="pre">MR()</span></code></a> call of this <a class="reference internal" href="modules.html#genal.Geno" title="genal.Geno"><code class="xref py py-class docutils literal notranslate"><span class="pre">Geno</span></code></a> instance.</p>
<p>If you wish to include the heterogeneity values (Cochran’s Q) in the results, you can use the heterogeneity argument in the <a class="reference internal" href="modules.html#id4" title="genal.Geno.MR"><code class="xref py py-meth docutils literal notranslate"><span class="pre">MR()</span></code></a> call. Here, the heterogeneity for the inverse-variance weighted method:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">SBP_clumped</span><span class="o">.</span><span class="n">MR</span><span class="p">(</span><span class="n">action</span><span class="o">=</span><span class="mi">2</span><span class="p">,</span> <span class="n">methods</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;Egger&quot;</span><span class="p">,</span><span class="s2">&quot;IVW&quot;</span><span class="p">],</span> <span class="n">exposure_name</span><span class="o">=</span><span class="s2">&quot;SBP&quot;</span><span class="p">,</span> <span class="n">outcome_name</span><span class="o">=</span><span class="s2">&quot;Stroke_eur&quot;</span><span class="p">,</span> <span class="n">heterogeneity</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
</pre></div>
</div>
<p>And that will give:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span>  <span class="n">exposure</span>     <span class="n">outcome</span>                      <span class="n">method</span>  <span class="n">nSNP</span>        <span class="n">b</span>       <span class="n">se</span>          <span class="n">pval</span>            <span class="n">Q</span>  <span class="n">Q_df</span>         <span class="n">Q_pval</span>
<span class="mi">0</span>      <span class="n">SBP</span>  <span class="n">Stroke_eur</span>                   <span class="n">MR</span> <span class="n">Egger</span>  <span class="mi">1499</span>  <span class="mf">0.027543</span>  <span class="mf">0.002849</span>  <span class="mf">1.723156e-21</span>  <span class="mf">2959.965136</span>  <span class="mi">1497</span>  <span class="mf">1.253763e-98</span>
<span class="mi">1</span>      <span class="n">SBP</span>  <span class="n">Stroke_eur</span>            <span class="n">Egger</span> <span class="n">Intercept</span>  <span class="mi">1499</span> <span class="o">-</span><span class="mf">0.001381</span>  <span class="mf">0.000813</span>  <span class="mf">8.935529e-02</span>  <span class="mf">2959.965136</span>  <span class="mi">1497</span>  <span class="mf">1.253763e-98</span>
<span class="mi">2</span>      <span class="n">SBP</span>  <span class="n">Stroke_eur</span>  <span class="n">Inverse</span><span class="o">-</span><span class="n">Variance</span> <span class="n">Weighted</span>  <span class="mi">1499</span>  <span class="mf">0.023049</span>  <span class="mf">0.001061</span>  <span class="mf">1.382645e-104</span> <span class="mf">2965.678836</span>  <span class="mi">1498</span>  <span class="mf">4.280737e-99</span>
</pre></div>
</div>
<p>As expected, many MR methods indicate that SBP is strongly associated with stroke, but there could be concerns for horizontal pleiotropy (instruments influencing the outcome through a different pathway than the one used as exposure) given the almost significant MR-Egger intercept p-value.</p>
<p>To investigate horizontal pleiotropy in more detail, a very useful method is Mendelian Randomization Pleiotropy RESidual Sum and Outlier (MR-PRESSO).
MR-PRESSO is a method designed to detect and correct for horizontal pleiotropy.
It will identify which instruments are likely to be pleiotropic on their effect on the outcome, and it will rerun an inverse-variance weighted MR after excluding them.
It can be run using the <a class="reference internal" href="modules.html#id5" title="genal.Geno.MRpresso"><code class="xref py py-meth docutils literal notranslate"><span class="pre">MRpresso()</span></code></a> method:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">mod_table</span><span class="p">,</span> <span class="n">GlobalTest</span><span class="p">,</span> <span class="n">OutlierTest</span><span class="p">,</span> <span class="n">BiasTest</span> <span class="o">=</span> <span class="n">SBP_clumped</span><span class="o">.</span><span class="n">MRpresso</span><span class="p">(</span><span class="n">action</span><span class="o">=</span><span class="mi">2</span><span class="p">,</span> <span class="n">n_iterations</span><span class="o">=</span><span class="mi">30000</span><span class="p">)</span>
</pre></div>
</div>
<p>As with the <a class="reference internal" href="modules.html#id4" title="genal.Geno.MR"><code class="xref py py-meth docutils literal notranslate"><span class="pre">MR()</span></code></a> method, the <code class="docutils literal notranslate"><span class="pre">action</span></code> argument determines how the pleiotropic SNPs will be treated. The output is a list containing:</p>
<ul class="simple">
<li><p>A table containing the original and outlier-corrected inverse variance-weighted results.</p></li>
<li><p>The global test p-value indicating the presence of horizontal pleiotropy.</p></li>
<li><p>A dataframe of p-values, one for each instrument, representing the likelihood that this instrument is pleiotropic (only relevant if the global test is significant).</p></li>
<li><dl class="simple">
<dt>A dictionary containing the outputs of the distortion test. This test assesses whether the removal of the pleiotropic instruments has significantly altered the original MR estimate.</dt><dd><ul>
<li><p>An array containing the indices of the pleiotropic SNPs.</p></li>
<li><p>The coefficient of the distortion test.</p></li>
<li><p>The p-value of the distortion test.</p></li>
</ul>
</dd>
</dl>
</li>
</ul>
</section>
<section id="snp-association-testing">
<h2>SNP-association testing<a class="headerlink" href="#snp-association-testing" title="Link to this heading">¶</a></h2>
<p>We may want to calibrate instrument-trait estimates in a specific population for which we have individual-level data (genetic files as well as phenotypic data). For instance, if the GWAS of SBP was done in a European population, we may want to adjust the estimates based on data coming from a population of a different ancestry. This can be done in 2 steps:</p>
<ol class="arabic simple">
<li><p>Loading the phenotypic data in a dataframe and calling the <a class="reference internal" href="modules.html#id6" title="genal.Geno.set_phenotype"><code class="xref py py-meth docutils literal notranslate"><span class="pre">set_phenotype()</span></code></a> method</p></li>
<li><p>Calling the <a class="reference internal" href="modules.html#id7" title="genal.Geno.association_test"><code class="xref py py-meth docutils literal notranslate"><span class="pre">association_test()</span></code></a> method to run the association tests and update the estimates</p></li>
</ol>
<p>Let’s start by loading phenotypic data:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">df_pheno</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="s2">&quot;path/to/trait/data&quot;</span><span class="p">)</span>
</pre></div>
</div>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>One important point is to make sure that the IDs of the participants are identical in the phenotypic data and in the genetic data.</p>
</div>
<p>Then, it is advised to make a copy of the <a class="reference internal" href="modules.html#genal.Geno" title="genal.Geno"><code class="xref py py-class docutils literal notranslate"><span class="pre">Geno</span></code></a> instance containing our instruments as we are going to update their coefficients and to avoid any confusion:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">SBP_adjusted</span> <span class="o">=</span> <span class="n">SBP_clumped</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
</pre></div>
</div>
<p>We can then call the <a class="reference internal" href="modules.html#id6" title="genal.Geno.set_phenotype"><code class="xref py py-meth docutils literal notranslate"><span class="pre">set_phenotype()</span></code></a> method, specifying which column contains our trait of interest (for the association testing) and which column contains the individual IDs:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">SBP_adjusted</span><span class="o">.</span><span class="n">set_phenotype</span><span class="p">(</span><span class="n">df_pheno</span><span class="p">,</span> <span class="n">PHENO</span><span class="o">=</span><span class="s2">&quot;htn&quot;</span><span class="p">,</span> <span class="n">IID</span><span class="o">=</span><span class="s2">&quot;IID&quot;</span><span class="p">)</span>
</pre></div>
</div>
<p>At this point, genal will identify if the phenotype is binary or quantitative in order to choose the appropriate regression model. If the phenotype is binary, it will assume that the most frequent value is coding for control (and the other value for case), this can be changed with <code class="docutils literal notranslate"><span class="pre">alternate_control=True</span></code>:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">Detected</span> <span class="n">a</span> <span class="n">binary</span> <span class="n">phenotype</span> <span class="ow">in</span> <span class="n">the</span> <span class="s1">&#39;PHENO&#39;</span> <span class="n">column</span><span class="o">.</span> <span class="n">Specify</span> <span class="s1">&#39;PHENO_type=&quot;quant&quot;&#39;</span> <span class="k">if</span> <span class="n">this</span> <span class="ow">is</span> <span class="n">incorrect</span><span class="o">.</span>
<span class="n">Identified</span> <span class="mi">0</span> <span class="k">as</span> <span class="n">the</span> <span class="n">control</span> <span class="n">code</span> <span class="ow">in</span> <span class="s1">&#39;PHENO&#39;</span><span class="o">.</span> <span class="n">Set</span> <span class="s1">&#39;alternate_control=True&#39;</span> <span class="n">to</span> <span class="n">inverse</span> <span class="n">this</span> <span class="n">interpretation</span><span class="o">.</span>
<span class="n">The</span> <span class="n">phenotype</span> <span class="n">data</span> <span class="ow">is</span> <span class="n">stored</span> <span class="ow">in</span> <span class="n">the</span> <span class="o">.</span><span class="n">phenotype</span> <span class="n">attribute</span><span class="o">.</span>
</pre></div>
</div>
<p>We can then run the association tests, specifying the path to the genetic files in plink format, and any columns we may want to include as covariates in the regression tests (making sure that the covariates are all numerical):</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">SBP_adjusted</span><span class="o">.</span><span class="n">association_test</span><span class="p">(</span><span class="n">covar</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;age&quot;</span><span class="p">],</span> <span class="n">path</span><span class="o">=</span><span class="s2">&quot;path/to/genetic/files&quot;</span><span class="p">)</span>
</pre></div>
</div>
<p>Genal will print information regarding the number of individuals used in the tests and the kind of tests performed. It is advised to make sure that this information is consistent with your data:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">CHR</span><span class="o">/</span><span class="n">POS</span> <span class="n">columns</span> <span class="n">present</span><span class="p">:</span> <span class="n">SNPs</span> <span class="n">searched</span> <span class="n">based</span> <span class="n">on</span> <span class="n">genomic</span> <span class="n">positions</span><span class="o">.</span>
<span class="n">Extracting</span> <span class="n">SNPs</span> <span class="k">for</span> <span class="n">each</span> <span class="n">chromosome</span><span class="o">...</span>
<span class="n">SNPs</span> <span class="n">extracted</span> <span class="k">for</span> <span class="n">chr1</span><span class="o">.</span>
<span class="n">SNPs</span> <span class="n">extracted</span> <span class="k">for</span> <span class="n">chr2</span><span class="o">.</span>
<span class="n">SNPs</span> <span class="n">extracted</span> <span class="k">for</span> <span class="n">chr3</span><span class="o">.</span>
<span class="n">SNPs</span> <span class="n">extracted</span> <span class="k">for</span> <span class="n">chr4</span><span class="o">.</span>
<span class="n">SNPs</span> <span class="n">extracted</span> <span class="k">for</span> <span class="n">chr5</span><span class="o">.</span>
<span class="n">SNPs</span> <span class="n">extracted</span> <span class="k">for</span> <span class="n">chr6</span><span class="o">.</span>
<span class="n">SNPs</span> <span class="n">extracted</span> <span class="k">for</span> <span class="n">chr7</span><span class="o">.</span>
<span class="n">SNPs</span> <span class="n">extracted</span> <span class="k">for</span> <span class="n">chr8</span><span class="o">.</span>
<span class="n">SNPs</span> <span class="n">extracted</span> <span class="k">for</span> <span class="n">chr9</span><span class="o">.</span>
<span class="n">SNPs</span> <span class="n">extracted</span> <span class="k">for</span> <span class="n">chr10</span><span class="o">.</span>
<span class="n">SNPs</span> <span class="n">extracted</span> <span class="k">for</span> <span class="n">chr11</span><span class="o">.</span>
<span class="n">SNPs</span> <span class="n">extracted</span> <span class="k">for</span> <span class="n">chr12</span><span class="o">.</span>
<span class="n">SNPs</span> <span class="n">extracted</span> <span class="k">for</span> <span class="n">chr13</span><span class="o">.</span>
<span class="n">SNPs</span> <span class="n">extracted</span> <span class="k">for</span> <span class="n">chr14</span><span class="o">.</span>
<span class="n">SNPs</span> <span class="n">extracted</span> <span class="k">for</span> <span class="n">chr15</span><span class="o">.</span>
<span class="n">SNPs</span> <span class="n">extracted</span> <span class="k">for</span> <span class="n">chr16</span><span class="o">.</span>
<span class="n">SNPs</span> <span class="n">extracted</span> <span class="k">for</span> <span class="n">chr17</span><span class="o">.</span>
<span class="n">SNPs</span> <span class="n">extracted</span> <span class="k">for</span> <span class="n">chr18</span><span class="o">.</span>
<span class="n">SNPs</span> <span class="n">extracted</span> <span class="k">for</span> <span class="n">chr19</span><span class="o">.</span>
<span class="n">SNPs</span> <span class="n">extracted</span> <span class="k">for</span> <span class="n">chr20</span><span class="o">.</span>
<span class="n">SNPs</span> <span class="n">extracted</span> <span class="k">for</span> <span class="n">chr21</span><span class="o">.</span>
<span class="n">SNPs</span> <span class="n">extracted</span> <span class="k">for</span> <span class="n">chr22</span><span class="o">.</span>
<span class="n">Merging</span> <span class="n">SNPs</span> <span class="n">extracted</span> <span class="kn">from</span> <span class="nn">each</span> <span class="n">chromosome</span><span class="o">...</span>
<span class="n">Created</span> <span class="n">bed</span><span class="o">/</span><span class="n">bim</span><span class="o">/</span><span class="n">fam</span> <span class="n">fileset</span> <span class="k">with</span> <span class="n">extracted</span> <span class="n">SNPs</span><span class="p">:</span> <span class="n">tmp_GENAL</span><span class="o">/</span><span class="n">e415aab3_allchr</span>
<span class="mi">39131</span> <span class="n">individuals</span> <span class="n">are</span> <span class="n">present</span> <span class="ow">in</span> <span class="n">the</span> <span class="n">genetic</span> <span class="n">data</span> <span class="ow">and</span> <span class="n">have</span> <span class="n">a</span> <span class="n">valid</span> <span class="n">phenotype</span> <span class="n">trait</span><span class="o">.</span>
<span class="n">Running</span> <span class="n">single</span><span class="o">-</span><span class="n">SNP</span> <span class="n">logistic</span> <span class="n">regression</span> <span class="n">tests</span> <span class="n">on</span> <span class="n">tmp_GENAL</span><span class="o">/</span><span class="n">e415aab3_allchr</span> <span class="n">data</span> <span class="k">with</span> <span class="n">adjustment</span> <span class="k">for</span><span class="p">:</span> <span class="p">[</span><span class="s1">&#39;age&#39;</span><span class="p">]</span><span class="o">.</span>
<span class="n">The</span> <span class="n">BETA</span><span class="p">,</span> <span class="n">SE</span><span class="p">,</span> <span class="n">P</span> <span class="n">columns</span> <span class="n">of</span> <span class="n">the</span> <span class="o">.</span><span class="n">data</span> <span class="n">attribute</span> <span class="n">have</span> <span class="n">been</span> <span class="n">updated</span><span class="o">.</span>
</pre></div>
</div>
<p>The <code class="docutils literal notranslate"><span class="pre">BETA</span></code>, <code class="docutils literal notranslate"><span class="pre">SE</span></code>, and <code class="docutils literal notranslate"><span class="pre">P</span></code> columns of the <code class="docutils literal notranslate"><span class="pre">SBP_adjusted.data</span></code> attribute have been updated with the results of the association tests.</p>
</section>
<section id="lifting">
<h2>Lifting<a class="headerlink" href="#lifting" title="Link to this heading">¶</a></h2>
<p>It is sometimes necessary to lift the SNP data to a different build. For instance, if the genetic data of our target population is in build 38 (hg38), but the GWAS summary statistics are in build 37 (hg19).
This can easily be done in genal using the <a class="reference internal" href="modules.html#id8" title="genal.Geno.lift"><code class="xref py py-meth docutils literal notranslate"><span class="pre">lift()</span></code></a> method:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">SBP_clumped</span><span class="o">.</span><span class="n">lift</span><span class="p">(</span><span class="n">start</span><span class="o">=</span><span class="s2">&quot;hg19&quot;</span><span class="p">,</span> <span class="n">end</span><span class="o">=</span><span class="s2">&quot;hg38&quot;</span><span class="p">,</span> <span class="n">replace</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
</pre></div>
</div>
<p>This outputs a table with the lifted SBP instruments (stored in the <code class="docutils literal notranslate"><span class="pre">SBP_clumped</span></code> instance) from build 37 (hg19) to build 38 (hg38). We specified <code class="docutils literal notranslate"><span class="pre">replace=False</span></code> to not modify the <code class="docutils literal notranslate"><span class="pre">SBP_clumped.data</span></code> attribute, but we may want to modify it (before running a PRS in a population stored in build 38 for instance).
Genal will download the appropriate chain files required for the lift, and it will be done in  python by default. However, if you plan to lift large datasets of SNPs (the whole summary statistics for instance), it may be useful to install the LiftOver executable that will run faster than the python version. It can be downloaded here: <a class="reference external" href="https://genome-store.ucsc.edu/">https://genome-store.ucsc.edu/</a> You will need to create an account, scroll down to “LiftOver program”, add it to your cart, and declare that you are a non-profit user.</p>
<p>You can specify the path of the LiftOver executable to the <code class="docutils literal notranslate"><span class="pre">liftover_path</span></code> argument:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">SBP_Geno</span><span class="o">.</span><span class="n">lift</span><span class="p">(</span><span class="n">start</span><span class="o">=</span><span class="s2">&quot;hg19&quot;</span><span class="p">,</span> <span class="n">end</span><span class="o">=</span><span class="s2">&quot;hg38&quot;</span><span class="p">,</span> <span class="n">replace</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">liftover_path</span><span class="o">=</span><span class="s2">&quot;path/to/liftover/exec&quot;</span><span class="p">)</span>
</pre></div>
</div>
</section>
<section id="gwas-catalog">
<h2>GWAS Catalog<a class="headerlink" href="#gwas-catalog" title="Link to this heading">¶</a></h2>
<p>It is sometimes interesting to determine the traits associated with our SNPs. In Mendelian Randomization, for instance, we may want to exclude instruments that are associated with traits likely causing horizontal pleiotropy. For this purpose, we can use the <a class="reference internal" href="modules.html#id9" title="genal.Geno.query_gwas_catalog"><code class="xref py py-meth docutils literal notranslate"><span class="pre">query_gwas_catalog()</span></code></a> method. This method will query the GWAS Catalog API to determine the list of traits associated with each of our SNPs and store the results in a list in the <code class="docutils literal notranslate"><span class="pre">ASSOC</span></code> column of the <code class="docutils literal notranslate"><span class="pre">.data</span></code> attribute:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">SBP_clumped</span><span class="o">.</span><span class="n">query_gwas_catalog</span><span class="p">(</span><span class="n">p_threshold</span><span class="o">=</span><span class="mf">5e-8</span><span class="p">)</span>
</pre></div>
</div>
<p>Which will output:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>Querying the GWAS Catalog and creating the ASSOC column.
Only associations with a p-value &lt;= 5e-08 are reported. Use the p_threshold argument to change the threshold.
To report the p-value for each association, use return_p=True.
To report the study ID for each association, use return_study=True.
The .data attribute will be modified. Use replace=False to leave it as is.
100%|██████████| 1545/1545 [00:34&lt;00:00, 44.86it/s]
The ASSOC column has been successfully created.
701 (45.37%) SNPs failed to query (not found in GWAS Catalog) and 7 (0.5%) SNPs timed out after 34.33 seconds. You can increase the timeout value with the timeout argument.
</pre></div>
</div>
<p>And the <a class="reference internal" href="modules.html#genal.Geno.data" title="genal.Geno.data"><code class="xref py py-attr docutils literal notranslate"><span class="pre">data</span></code></a> attribute now contains an <cite>ASSOC</cite> column:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">EA</span> <span class="n">NEA</span>    <span class="n">EAF</span>    <span class="n">BETA</span>     <span class="n">SE</span>  <span class="n">CHR</span>        <span class="n">POS</span>         <span class="n">SNP</span>                                               <span class="n">ASSOC</span>
<span class="mi">0</span>  <span class="n">A</span>   <span class="n">G</span>  <span class="mf">0.1784</span>  <span class="mf">0.2330</span>  <span class="mf">0.0402</span>   <span class="mi">10</span>  <span class="mi">102075479</span>    <span class="n">rs603424</span>  <span class="p">[</span><span class="n">eicosanoids</span> <span class="n">measurement</span><span class="p">,</span> <span class="n">decadienedioic</span> <span class="n">acid</span> <span class="p">(</span><span class="o">...</span><span class="p">]</span>
<span class="mi">1</span>  <span class="n">A</span>   <span class="n">G</span>  <span class="mf">0.0706</span> <span class="o">-</span><span class="mf">0.3873</span>  <span class="mf">0.0626</span>   <span class="mi">10</span>  <span class="mi">102403682</span>   <span class="n">rs2996303</span>                                       <span class="n">FAILED_QUERY</span>
<span class="mi">2</span>  <span class="n">T</span>   <span class="n">G</span>  <span class="mf">0.8872</span>  <span class="mf">0.6846</span>  <span class="mf">0.0480</span>   <span class="mi">10</span>  <span class="mi">102553647</span>   <span class="n">rs1006545</span>  <span class="p">[</span><span class="n">diastolic</span> <span class="n">blood</span> <span class="n">pressure</span><span class="p">,</span> <span class="n">systolic</span> <span class="n">blood</span> <span class="n">pressure</span><span class="o">...</span><span class="p">]</span>
<span class="mi">3</span>  <span class="n">T</span>   <span class="n">G</span>  <span class="mf">0.6652</span> <span class="o">-</span><span class="mf">0.2098</span>  <span class="mf">0.0340</span>   <span class="mi">10</span>  <span class="mi">102558506</span>  <span class="n">rs12570050</span>                                       <span class="n">FAILED_QUERY</span>
<span class="mi">4</span>  <span class="n">T</span>   <span class="n">C</span>  <span class="mf">0.3057</span> <span class="o">-</span><span class="mf">0.2448</span>  <span class="mf">0.0334</span>   <span class="mi">10</span>  <span class="mi">102603924</span>   <span class="n">rs4919502</span>                                       <span class="n">FAILED_QUERY</span>
<span class="mi">5</span>  <span class="o">...</span> <span class="o">...</span>    <span class="o">...</span>    <span class="o">...</span>    <span class="o">...</span>  <span class="o">...</span>        <span class="o">...</span>         <span class="o">...</span>                                                <span class="o">...</span>
<span class="mi">6</span>  <span class="n">T</span>   <span class="n">C</span>  <span class="mf">0.3514</span>  <span class="mf">0.2203</span>  <span class="mf">0.0314</span>    <span class="mi">9</span>   <span class="mi">9350706</span>    <span class="n">rs1332813</span>  <span class="p">[</span><span class="n">diastolic</span> <span class="n">blood</span> <span class="n">pressure</span><span class="p">,</span> <span class="n">systolic</span> <span class="n">blood</span> <span class="n">pressure</span><span class="o">...</span><span class="p">]</span>
<span class="mi">7</span>  <span class="n">T</span>   <span class="n">C</span>  <span class="mf">0.6880</span> <span class="o">-</span><span class="mf">0.1897</span>  <span class="mf">0.0332</span>    <span class="mi">9</span>  <span class="mi">94201341</span>  <span class="n">rs10820855</span>                                       <span class="n">FAILED_QUERY</span>
<span class="mi">8</span>  <span class="n">A</span>   <span class="n">T</span>  <span class="mf">0.3669</span> <span class="o">-</span><span class="mf">0.1862</span>  <span class="mf">0.0313</span>    <span class="mi">9</span>  <span class="mi">95201540</span>   <span class="n">rs7045409</span>  <span class="p">[</span><span class="n">protein</span> <span class="n">measurement</span><span class="p">,</span> <span class="n">pulse</span> <span class="n">pressure</span> <span class="n">measurement</span><span class="o">...</span><span class="p">]</span>
</pre></div>
</div>
<p>If you are also interested in the p-values of each SNP-trait association, or the ID of the study from which the association was reported, you can use the <code class="docutils literal notranslate"><span class="pre">return_p</span> <span class="pre">=</span> <span class="pre">True</span></code> and <code class="docutils literal notranslate"><span class="pre">return_study</span> <span class="pre">=</span> <span class="pre">True</span></code> arguments. Then, the <code class="docutils literal notranslate"><span class="pre">ASSOC</span></code> column will contain a list of tuples, where each tuple contains the trait name, the p-value, and the study ID:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">SBP_clumped</span><span class="o">.</span><span class="n">query_gwas_catalog</span><span class="p">(</span><span class="n">p_threshold</span><span class="o">=</span><span class="mf">5e-8</span><span class="p">,</span> <span class="n">return_p</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">return_study</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
</pre></div>
</div>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>  <span class="n">EA</span> <span class="n">NEA</span>    <span class="n">EAF</span>    <span class="n">BETA</span>     <span class="n">SE</span>  <span class="n">CHR</span>        <span class="n">POS</span>         <span class="n">SNP</span>                                               <span class="n">ASSOC</span>
<span class="mi">0</span>  <span class="n">A</span>   <span class="n">G</span>  <span class="mf">0.1784</span>  <span class="mf">0.2330</span>  <span class="mf">0.0402</span>   <span class="mi">10</span>  <span class="mi">102075479</span>    <span class="n">rs603424</span>                                            <span class="n">TIMEOUT</span>
<span class="mi">1</span>  <span class="n">A</span>   <span class="n">G</span>  <span class="mf">0.0706</span> <span class="o">-</span><span class="mf">0.3873</span>  <span class="mf">0.0626</span>   <span class="mi">10</span>  <span class="mi">102403682</span>   <span class="n">rs2996303</span>                                       <span class="n">FAILED_QUERY</span>
<span class="mi">2</span>  <span class="n">T</span>   <span class="n">G</span>  <span class="mf">0.8872</span>  <span class="mf">0.6846</span>  <span class="mf">0.0480</span>   <span class="mi">10</span>  <span class="mi">102553647</span>   <span class="n">rs1006545</span>  <span class="p">[(</span><span class="n">heart</span> <span class="n">rate</span> <span class="n">response</span> <span class="n">to</span> <span class="n">exercise</span><span class="p">,</span> <span class="mf">6e-12</span><span class="p">,</span> <span class="n">GCST</span><span class="o">...</span>
<span class="mi">3</span>  <span class="n">T</span>   <span class="n">G</span>  <span class="mf">0.6652</span> <span class="o">-</span><span class="mf">0.2098</span>  <span class="mf">0.0340</span>   <span class="mi">10</span>  <span class="mi">102558506</span>  <span class="n">rs12570050</span>                                       <span class="n">FAILED_QUERY</span>
<span class="mi">4</span>  <span class="n">T</span>   <span class="n">C</span>  <span class="mf">0.3057</span> <span class="o">-</span><span class="mf">0.2448</span>  <span class="mf">0.0334</span>   <span class="mi">10</span>  <span class="mi">102603924</span>   <span class="n">rs4919502</span>                                       <span class="n">FAILED_QUERY</span>
<span class="mi">5</span>  <span class="o">...</span> <span class="o">...</span>    <span class="o">...</span>    <span class="o">...</span>    <span class="o">...</span>  <span class="o">...</span>        <span class="o">...</span>         <span class="o">...</span>                                                <span class="o">...</span>
<span class="mi">6</span>  <span class="n">T</span>   <span class="n">C</span>  <span class="mf">0.3514</span>  <span class="mf">0.2203</span>  <span class="mf">0.0314</span>    <span class="mi">9</span>   <span class="mi">9350706</span>    <span class="n">rs1332813</span>  <span class="p">[(</span><span class="n">diastolic</span> <span class="n">blood</span> <span class="n">pressure</span><span class="p">,</span> <span class="mf">1e-12</span><span class="p">,</span> <span class="n">GCST9031029</span><span class="o">...</span>
<span class="mi">7</span>  <span class="n">T</span>   <span class="n">C</span>  <span class="mf">0.6880</span> <span class="o">-</span><span class="mf">0.1897</span>  <span class="mf">0.0332</span>    <span class="mi">9</span>  <span class="mi">94201341</span>  <span class="n">rs10820855</span>                                       <span class="n">FAILED_QUERY</span>
<span class="mi">8</span>  <span class="n">A</span>   <span class="n">T</span>  <span class="mf">0.3669</span> <span class="o">-</span><span class="mf">0.1862</span>  <span class="mf">0.0313</span>    <span class="mi">9</span>  <span class="mi">95201540</span>   <span class="n">rs7045409</span>  <span class="p">[(</span><span class="n">systolic</span> <span class="n">blood</span> <span class="n">pressure</span><span class="p">,</span> <span class="mf">9e-13</span><span class="p">,</span> <span class="n">GCST006624</span><span class="p">),</span><span class="o">...</span>
</pre></div>
</div>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>As you can see, many SNPs failed to be queried. This is normal as the GWAS Catalog is not exhaustive.</p>
</div>
</section>
</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="index.html" class="btn btn-neutral float-left" title="genal: A Python Toolkit for Genetic Risk Scoring and Mendelian Randomization" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="modules.html" class="btn btn-neutral float-right" title="The Geno class" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2023, Cyprien A. Rivier.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>